[TP] Enable more generic attn in Tensor Parallelism

fduwjj opened the pull request:
To make TP more generic for Attention module, we come up with this new col/rowwise parallel style.

Basically, the idea behind is that:
We only do DTensor op for Col/Rowwise sharded part. For the rest of ATen ops, we will leave it to Tensor ops.

And we set this behavior as default for Colwise and Rowwise parallel style. If people want to customize it, they can always pass in different prepare_input or prepare_output


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100508
* #100137

Commit:
Enable more generic attn in Tensor Parallelism

[ghstack-poisoned]

Commit:
Update on "Enable more generic attn in Tensor Parallelism"




[ghstack-poisoned]

wanchaol commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -388,3 +325,67 @@ def make_output_reshard_tensor(
+        super().__init__(None, make_output_replicate_1d)
+
+
+class ColwiseParallelForAttn(ColwiseParallel):
```
Can we just merge this to `ColwiseParallel` itself? Maybe add a init arg if you want to distinguish difference of input/output control

wanchaol commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -388,3 +325,67 @@ def make_output_reshard_tensor(
+        super().__init__(make_input_replicate_1d, make_sharded_output_tensor)
+
+
+class RowwiseParallelForAttn(RowwiseParallel):
```
Let's try to make sure we only have `RowwiseParallel` and `ColwiseParallel`, and `PairwiseParallel`, I feel adding a new style with subtle difference on input/output might not be a good UX.

wanchaol reviewed (commented):
First pass on this PR, mainly I would like to see if we can consolidate those different styles. imo exposing different style variants could really confuse the user if the style is not quite distinguishable from others.

Commit:
Update on "Enable more generic attn in Tensor Parallelism"

To make TP more generic for Attention module, we come up with this new col/rowwise parallel style.

Basically, the idea behind is that:
We only do DTensor op for Col/Rowwise sharded part. For the rest of ATen ops, we will leave it to Tensor.





[ghstack-poisoned]

Commit:
Update on "[TP] Enable more generic attn in Tensor Parallelism"

To make TP more generic for Attention module, we come up with this new col/rowwise parallel style.

Basically, the idea behind is that:
We only do DTensor op for Col/Rowwise sharded part. For the rest of ATen ops, we will leave it to Tensor ops.

And we set this behavior as default for Colwise and Rowwise parallel style. If people want to customize it, they can always pass in different prepare_input or prepare_output





[ghstack-poisoned]

wanchaol commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -66,7 +65,7 @@ def __init__(self, _prepare_input=None, _prepare_output=None) -> None:
-class PairwiseSequenceParallel(PairwiseParallel):
+class SequenceParallel(PairwiseParallel):
     """
     PairwiseSequenceParallel concatenate colwise and rowwise styles as a fixed
```
can we also fix the docs so to the correct name?

fduwjj replied:
oh sure.

wanchaol commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -320,6 +277,28 @@ def make_output_tensor(
 
+@_prepare_output_validate  # type: ignore[arg-type] # pyre-ignore[56]
+def make_sharded_output_tensor(
+    output: DTensor, _device_mesh: Optional[DeviceMesh] = None
```
is there a reason we pass in `_device_mesh`? I thought we don't need this to convert a DTensor to torch.Tensor?

fduwjj replied:
right, but we do explicitly pass in here:
https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/api.py#L531

wanchaol commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -320,6 +277,28 @@ def make_output_tensor(
+    output: DTensor, _device_mesh: Optional[DeviceMesh] = None
+) -> torch.Tensor:
+    """
+    Convert sharded Output DTensor to Tensor.
```
nit: to `torch.Tensor`

wanchaol commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -320,6 +277,28 @@ def make_output_tensor(
+    Args:
+        output (:class:`DTensor`):
+            Output of module to be converted.
+        _device_mesh (:class:`DeviceMesh`, optional):
```
Oh I see, maybe we don't really need to document this argument? Added a comment to explain this should be enough

wanchaol commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -342,3 +321,23 @@ def make_output_reshard_tensor(
+class RowwiseParallel(ParallelStyle):
+    """
+    Partitioning the row of a module.
+    We assume the input to be a sharded :class:`DTensor` and output to be a replicated :class:`DTensor`.
```
nit: we should update the doc to reflect the fact it's returning a torch.Tensor

wanchaol commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -342,3 +321,23 @@ def make_output_reshard_tensor(
+    """
+    Partitioning the column of a tensor or module.
+    We assume the input to be a replicated :class:`DTensor` and output to be a sharded :class:`DTensor`.
+    """
```
ditto: sharded torch.Tensor

wanchaol reviewed (approved):
lgtm, we should update the docs to reflect the changes before landing

wanchaol reviewed (approved):
lgtm, we should update the docs to reflect the changes before landing

Commit:
Update on "[TP] Enable more generic attn in Tensor Parallelism"

To make TP more generic for Attention module, we come up with this new col/rowwise parallel style.

Basically, the idea behind is that:
We only do DTensor op for Col/Rowwise sharded part. For the rest of ATen ops, we will leave it to Tensor ops.

And we set this behavior as default for Colwise and Rowwise parallel style. If people want to customize it, they can always pass in different prepare_input or prepare_output





[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge
